rpc/database.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 Athanor Labs (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package rpc
+
+import (
+	"math/big"
+	"net/http"
+
+	"github.com/athanorlabs/atomic-swap/common/types"
+	mcrypto "github.com/athanorlabs/atomic-swap/crypto/monero"
+	"github.com/athanorlabs/atomic-swap/db"
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+// RecoveryDB contains methods for retrieving swap recovery info from the database.
+type RecoveryDB interface {
+	GetContractSwapInfo(id types.Hash) (*db.EthereumSwapInfo, error)
+	GetSwapPrivateKey(id types.Hash) (*mcrypto.PrivateSpendKey, error)
+	GetCounterpartySwapPrivateKey(id types.Hash) (*mcrypto.PrivateSpendKey, error)
+}
+
+// DatabaseService ...
+type DatabaseService struct {
+	rdb RecoveryDB
+}
+
+// NewDatabaseService returns a new DatabaseService.
+func NewDatabaseService(rdb RecoveryDB) *DatabaseService {
+	return &DatabaseService{
+		rdb: rdb,
+	}
+}
+
+// GetContractSwapInfoRequest ...
+type GetContractSwapInfoRequest struct {
+	ID types.Hash `json:"id" validate:"required"`
+}
+
+// GetContractSwapInfoResponse ...
+type GetContractSwapInfoResponse struct {
+	StartNumber     *big.Int                   `json:"startNumber" validate:"required"`
+	SwapID          types.Hash                 `json:"swapID" validate:"required"`
+	Swap            *contracts.SwapCreatorSwap `json:"swap" validate:"required"`
+	SwapCreatorAddr ethcommon.Address          `json:"swapCreatorAddr" validate:"required"`
+}
+
+// GetContractSwapInfo returns the contract swap info for the given swap ID from the database.
+func (s *DatabaseService) GetContractSwapInfo(
+	_ *http.Request,
+	req *GetContractSwapInfoRequest,
+	resp *GetContractSwapInfoResponse,
+) error {
+	info, err := s.rdb.GetContractSwapInfo(req.ID)
+	if err != nil {
+		return err
+	}
+
+	resp.StartNumber = info.StartNumber
+	resp.SwapID = info.SwapID
+	resp.Swap = info.Swap
+	resp.SwapCreatorAddr = info.SwapCreatorAddr
+	return nil
+}
+
+// GetSwapPrivateKeyRequest ...
+type GetSwapPrivateKeyRequest struct {
+	ID types.Hash `json:"id" validate:"required"`
+}
+
+// GetSwapPrivateKeyResponse ...
+type GetSwapPrivateKeyResponse struct {
+	PrivateKey *mcrypto.PrivateSpendKey `json:"privateKey" validate:"required"`
```

Comment:
Given that Monero has 2 private keys and the spend key isn't always derived from the view key, can we be precise of the field name?

Reply:
maybe this should just be called `SwapSecret` or something since it's never actually used as a monero spend key?

Reply:
Based on your response and studying the surrounding code, I now understand that this is "our" half of the private spend key. Here's a way you could blend your idea of using the phrase `SwapSecret` and be obvious as to whether the secret is ours/theirs/combined/view/spend:

```
type GetOurSwapSecretRequest struct {
	OfferID types.Hash `json:"offerID" validate:"required"`
}

type GetOurSwapSecretResponse struct {
	PrivSpendKey *mcrypto.PrivateSpendKey `json:"secret" validate:"required"`
}

type GetTheirSwapSecretRequest struct {
	OfferID types.Hash `json:"offerID" validate:"required"`
}

type GetTheirSwapSecretResponse struct {
	PrivSpendKey *mcrypto.PrivateSpendKey `json:"PrivSpendKey" validate:"required"`
}
```
I don't have anything against the word "counterparty", but these names are already on the long side and "counterparty" doesn't have a nice symmetric antonym.

Reply:
updated to be `GetSwapSecret`, I removed the counterparty secret functions since they aren't actually usable by the user to recover